src/transformStateWithClones.js

```diff
@@ -7,7 +7,30 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+  let previousState = { ...state };
+
+  for (const action of actions) {
+    let currentState = { ...previousState };
+
+    if (action.type === 'addProperties') {
+      Object.assign(currentState, action.extraData);
+    }
+
+    if (action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete currentState[key];
+      }
+    }
+
+    if (action.type === 'clear') {
+      currentState = {};
+    }
```

Comment:
Use `switch/case` structure instead of multiple `if`s against the same variable with well defined values; (also, in situations such as these, adding a `default` fallback option for when the supplied `type` is invalid is a good practice)